Reverts backwards incompatible changes to theano_function()

moorepants opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

Fixes #14986

#### Brief description of what is fixed or changed

PR #14949 introduced a backwards incompatible change to the public API
of the theano_function() function. The change allowed the output
expressions to be passed in without being encased in a sequence, but did
not retain the prior behavior. The expressed reasons for the change were
to allow a user to type `my_expr` instead of `[my_expr]` and to have a
similar call signature with `theano.function()`. Neither of these
warrant a necessary deprecation. In particular, commit
a833af605791c38518d133fd98d7d0e99190ff97 is reverted.

Secondly, this change inadvertently broke (seemingly) untested behavior
if the output expressions are matrices. Prior to the change a numpy
array of the same dimensions of the matrix would be returned. The change
caused lists of arrays to be returned. This returns that behavior and
adds a regression test.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->
(NOTE: The release notes for this PR need to be fixed manually once this is merged, see https://github.com/sympy/sympy/pull/15423#issuecomment-432519197)

moorepants committed:
Reverts backwards incompatible changes to theano_function()

PR #14949 introduced a backwards incompatible change to the public API
of the theano_function() function. The change allowed the output
expressions to be passed in without being encased in a sequence, but did
not retain the prior behavior. The expressed reasons for the change were
to allow a user to type `my_expr` instead of `[my_expr]` and to have a
similar call signature with `theano.function()`. Neither of these
warrant a necessary deprecation. In particular, commit
a833af605791c38518d133fd98d7d0e99190ff97 is reverted.

Secondly, this change inadvertently broke (seemingly) untested behavior
if the output expressions are matrices. Prior to the change a numpy
array of the same dimensions of the matrix would be returned. The change
caused lists of arrays to be returned. This returns that behavior and
adds a regression test.

moorepants committed:
Added one more regression test for matrices.

moorepants commented:
@jlumpe FYI, I am requesting to revert part of your PR here.

asmeurer commented:
That PR was merged since the last release, no? The release notes entry here should just be `NO ENTRY` and you should manually edit the release notes to remove the entry added from #14949.

asmeurer commented:
Let's wait a bit to see if @jlumpe has any comments.

moorepants commented:
I'll fix up the release notes manually once merged. This change is not in 1.3.

jlumpe commented:
Confirmed that changing line 514 to

    is_seq = not isinstance(outputs, (sympy.Basic, sympy.MatrixBase))

Turns out matrices don't actually inherit from `Expr` or even `Basic`. Not sure if there's anything else I should put in there, using `Basic` instead of `Expr` covers the case of `Array` objects as well. Any other sequence-like Sympy objects we should write tests for?

jlumpe commented:
@moorepants Sorry, that wasn't clear. I was referring to the `test_theano_function_matrix` test you added.

moorepants commented:
PR #14949 made several changes:

1. Changed the public API such that the `outputs` argument expects an expression or a sequence of expressions instead of just a sequence of expressions.
2. Changed existing tests, leaving the original API untested.
3. Added a temporary kwarg `squeeze` only for the purposes of dealing with deprecating the `outputs` API, which itself was also deprecated as soon as it was committed.
4. Added the new `scalar` kwarg and tests for it.
5. Inadvertently introduced a bug if `outputs` is a matrix. (This was not tested in SymPy but was in an upstream library).

I do not agree that any of the above changes should have been merged except for 4. This PR reverts 1, 2, 3, and 5. SymPy's public API should only be deprecated if absolutely necessary. In this case it is not necessary. I would like to merge this PR as is and if you want to propose a change that doesn't involve deprecating the API, I'm happy to review it.

jlumpe commented:
@moorepants I can understand any arguments about not deprecating the old behavior, but I disagree with reverting 1 and 3.

There might not be many problems when using the function with a hard-coded list of outputs, but anytime the list can be variable in length having a totally different pattern of behavior when the length is one is probably going to cause issues. I've run into the same problem with matplotlib's ``subplots()`` function many times before, which also has an optional ``squeeze=False`` argument to prevent exactly this.

Accepting a single expression instead of a list may not be completely necessary, but it shouldn't break backwards compatibility in any way. It matches up with the Theano API and I think is less confusing.

jlumpe commented:
I'm not advocating against removing the deprecation (I think you've made a good case against it), but just leaving the new feature (`squeeze` argument) in.

moorepants commented:
I may have misunderstood. I thought the deprecation warning indicated that `squeeze` was deprecated on arrival. The deprecation warnings I was receiving seemed to indicate this.

asmeurer commented:
What is the status of this?

moorepants commented:
The status is that I'd like to merge this because it fixes a break in the code. New work can be done to add new features, etc, after merged, IMHO.

jlumpe commented:
I see this still hasn't been merged. @asmeurer I think moorepants makes a good case for why this PR is a good idea. I would like to submit my own PR afterwards to re-introduce the `squeeze` kwarg in a way that doesn't affect backwards compatibility or deprecate anything.

asmeurer commented:
OK. Can one of you fix the release notes page. Also assumedly #14986 should be closed.
